src/map.c

```diff
@@ -3,752 +3,641 @@
+ * invaluable work.
+ * Reference:
+ *   https://github.com/jemalloc/jemalloc/blob/dev/include/jemalloc/internal/rb.h
+ */
+
 #include <assert.h>
 #include <stdint.h>
 #include <stdlib.h>
 #include <string.h>
 
 #include "map.h"
 
-struct map_internal {
-    struct map_node *head;
+/* TODO: Avoid relying on key_size and data_size */
+struct map_head_t {
+    map_node_t *root;
 
-    /* Properties */
-    size_t key_size, element_size, size;
+    /* properties */
+    size_t key_size, data_size;
 
-    map_iter_t it_end, it_most, it_least;
-
-    int (*comparator)(const void *, const void *);
+    map_cmp_t (*cmp)(const void *, const void *);
 };
 
-typedef enum { RB_RED = 0, RB_BLACK } map_color_t;
-
-/*
- * Get parent of node
- * @node: pointer to the rb node
- *
- * Return: rb parent node of @node
+/* Each node in the red-black tree consumes at least 1 byte of space (for the
+ * linkage if nothing else), so there are a maximum of sizeof(void *) << 3
+ * red-black tree nodes in any process (and thus, at most sizeof(void *) << 3
+ * nodes in any red-black tree). The choice of algorithm bounds the depth of
+ * a tree to twice the binary logarithm (base 2) of the number of elements in
+ * the tree; the following bound applies.
  */
-static inline map_node_t *rb_parent(const map_node_t *node)
-{
-    return (map_node_t *) (node->parent_color & ~1LU);
-}
+#define RB_MAX_DEPTH (sizeof(void *) << 4)
 
-/*
- * Get color of node
- * @node: pointer to the rb node
- *
- * Return: color of @node
- */
-static inline map_color_t rb_color(const map_node_t *node)
-{
-    return (map_color_t) (node->parent_color & 1LU);
-}
+typedef enum { RB_BLACK = 0, RB_RED } map_color_t;
 
-/*
- * Set parent of node
- * @node: pointer to the rb node
- * @parent: pointer to the new parent node
- */
-static inline void rb_set_parent(map_node_t *node, map_node_t *parent)
+/* Left accessors */
+static inline map_node_t *rb_node_get_left(const map_node_t *node)
 {
-    node->parent_color = (unsigned long) parent | (node->parent_color & 1LU);
+    return node->left;
 }
 
-/*
- * Set color of node
- * @node: pointer to the rb node
- * @color: new color of the node
- */
-static inline void rb_set_color(map_node_t *node, map_color_t color)
+static inline void rb_node_set_left(map_node_t *node, map_node_t *left)
 {
-    node->parent_color = (node->parent_color & ~1LU) | color;
+    node->left = left;
 }
 
-/*
- * Set parent and color of node
- * @node: pointer to the rb node
- * @parent: pointer to the new parent node
- * @color: new color of the node
- */
-static inline void rb_set_parent_color(map_node_t *node,
-                                       map_node_t *parent,
-                                       map_color_t color)
+/* Right accessors */
+static inline map_node_t *rb_node_get_right(const map_node_t *node)
 {
-    node->parent_color = (unsigned long) parent | color;
+    return (map_node_t *) (((uintptr_t) node->right_red) & ~3);
```

Comment:
Evaluate https://github.com/torvalds/linux/commit/b0687c1119b4e8c88a651b6e876b7eae28d076e3 and integrate into this pull request (as another commit).

Reply:
I do not think it is a good idea.

In linux the usage `rb_set_black` is only used to set node from red to black.
However, the map `rb_node_set_red` we implemented can set node from any state to black.
If we change the interface, we will cause a lot of unneeded branch that hurts the performance.